p8e-api/src/main/kotlin/io/provenance/engine/stream/EventStream.kt

```diff
@@ -0,0 +1,288 @@
+
+        ABCIApplicationGrpc.newBlockingStub(channel)
+    }
+
+    fun getStream(eventTypes: List<String>, startHeight: Long, observer: EventStreamResponseObserver<EventBatch>) = EventStream(URI(eventStreamProperties.websocketUri), eventTypes, startHeight, observer, transactionQueryService)
+
+    class EventStream(
+        val node: URI,
+        val eventTypes: List<String>,
+        val startHeight: Long,
+        private val observer: EventStreamResponseObserver<EventBatch>,
+        private val transactionQueryService: TransactionQueryService
+    ) {
+        companion object {
+            private val HISTORY_BATCH_SIZE = 10
+            private val executor = ThreadPoolFactory.newFixedThreadPool(HISTORY_BATCH_SIZE, "event-stream-%d")
+        }
+
+        private val log = logger()
+
+        private val lifecycle = LifecycleRegistry(0L)
+        private val scarletInstance = Scarlet.Builder()
+            .lifecycle(lifecycle)
+            .webSocketFactory(OkHttpClient.Builder()
+                .readTimeout(Duration.ofSeconds(60)) // ~ 12 pbc block cuts
+                .build()
+                .newWebSocketFactory("${node.scheme}://${node.host}:${node.port}/websocket"))
+            .addMessageAdapterFactory(MoshiMessageAdapter.Factory())
+            .addStreamAdapterFactory(RxJava2StreamAdapterFactory())
+            .build()
+        private val eventStreamService = scarletInstance.create<EventStreamService>()
+
+        private var subscription: Disposable? = null
+
+        private var lastBlockSeen = AtomicLong(-1)
+        private val eventMonitor = thread(start = false) {
+            var lastBlockMonitored = lastBlockSeen.get()
+            while (true) {
+                Thread.sleep(30000)
+                val lastBlockSeen = lastBlockSeen.get()
+                log.debug("Checking for event stream liveliness [lastBlockSeen: $lastBlockSeen vs. lastBlockMonitored: $lastBlockMonitored]")
+                if (lastBlockSeen <= lastBlockMonitored) {
+                    handleError(EventStreamStaleException("EventStream has not received a block in 30 seconds [lastBlockSeen: $lastBlockSeen vs. lastBlockMonitored: $lastBlockMonitored]"))
+                    break
+                }
+                lastBlockMonitored = lastBlockSeen
+            }
+            log.info("Exiting event monitor thread")
+        }
+
+        fun streamEvents() {
+            // todo: need to limit how many times this function is called??? Used to limit based on consumer id... probably need to use redis to do this... and ensure cleaned up when shutting down if do
+
+            // start event loop to start listening for events
+            startEventLoop()
+            shutdownHook { shutdown() } // ensure we close socket gracefully when shutting down
+
+            timed("EventStream:streamHistory") {
+                streamHistory()
+            }
+
+            eventMonitor.start()
+        }
+
+        private fun streamHistory() {
```

Comment:
Does fetching the history here basically perform the same function as the tx reaper we have that would process skipped events?

Reply:
um, it is a bit different, this is streaming history from the last block we recorded to the current latest block, while the error reaper is a fallback to process blocks if we somehow skip over them and the event stream has continued on... it could be that now that we have the event stream in-house this could be changed around a bit, but for now they are doing separate things

Reply:
👍